query-engine/query-engine-c-abi/src/engine.rs

```diff
@@ -0,0 +1,674 @@
+        // .with_subscriber(dispatcher)
+        // .await
+    }
+
+    // If connected, attempts to roll back a transaction with id `tx_id` in the core.
+    pub async fn rollback_transaction(&self, tx_id_str: *const c_char, _trace: *const c_char) -> Result<String, Error> {
+        let tx_id = get_cstr_safe(tx_id_str).expect("Input string missing");
+        // let trace = get_cstr_safe(trace_str).expect("trace is required in transactions");
+        let inner = self.inner.read().await;
+        let engine = inner.as_engine()?;
+
+        // let dispatcher = self.logger.dispatcher();
+
+        // async move {
+        match engine.executor().rollback_tx(TxId::from(tx_id)).await {
+            Ok(_) => Ok("{}".to_string()),
+            Err(err) => Ok(map_known_error(err)?),
+        }
+        // }
+        // .with_subscriber(dispatcher)
+        // .await
+    }
+}
+
+//            _____ _____
+//      /\   |  __ \_   _|
+//     /  \  | |__) || |
+//    / /\ \ |  ___/ | |
+//   / ____ \| |    _| |_
+//  /_/    \_\_|   |_____|
+//
+// This API is meant to be stateless. This means the box pointer to the query engine structure will be returned to the
+// calling code and should be passed to subsequent calls
+//
+// Be should be careful about not de-allocating the pointer
+// when adding a new function remember to always call mem::forget
+
+/// # Safety
+///
+/// The calling context needs to pass a valid pointer that will store the reference
+#[no_mangle]
+pub unsafe extern "C" fn prisma_create(options: ConstructorOptions, qe_ptr: *mut *mut QueryEngine) -> c_int {
+    if qe_ptr.is_null() {
+        return PRISMA_MISSING_POINTER;
+    }
+
+    let res = QueryEngine::new(options);
+    match res {
+        Ok(v) => {
+            *qe_ptr = Box::into_raw(Box::new(v));
+            PRISMA_OK
+        }
+        Err(e) => {
+            println!("Error creating query engine {:#?}", e);
+            PRISMA_UNKNOWN_ERROR
+        }
+    }
+}
+
+/// # Safety
+///
+/// The calling context needs to pass a valid pointer that will store the reference to the error string
+#[no_mangle]
+pub unsafe extern "C" fn prisma_connect(
+    qe: *mut QueryEngine,
+    trace: *const c_char,
+    error_string_ptr: *mut *mut c_char,
+) -> c_int {
+    let query_engine: Box<QueryEngine> = Box::from_raw(qe);
+    let result = RUNTIME.block_on(async { query_engine.connect(trace).await });
+
+    match result {
+        Ok(_engine) => {
+            std::mem::forget(query_engine);
+            *error_string_ptr = std::ptr::null_mut();
+            PRISMA_OK
+        }
+        Err(err) => {
+            RUNTIME.block_on(async {
+                let mut error_string = query_engine.error_string.write().await;
+                *error_string = CString::new(err.message()).unwrap();
+                *error_string_ptr = error_string.as_ptr() as *mut c_char;
+            });
+            std::mem::forget(query_engine);
+            PRISMA_UNKNOWN_ERROR
+        }
+    }
+}
```

Comment:
I think I did that, but as soon as the Rust fn finished the memory was cleaned, so the C calling function trying to print the string would get a corrupted heap error. Let me give it another try though.

Reply:
@ospfranco that's because `CString::as_ptr` is incredibly unsafe and easy to misuse 😄 If you did it like in @Weakky's snippet above, then the pointer, in fact, would become dangling even before exiting this function (if there was more code in it, here it's basically at the end anyway) but the moment a temporary is dropped at the end of the expression since `CString::new(err.message()).unwrap()` is not assigned to any variable. The correct way which will work is using `into_raw()` instead of `as_ptr()`:

```rust
            *error_string_ptr = CString::new(err.message()).unwrap().into_raw() as *mut c_char;
```

It will consume the `CString` and give up ownership of data. It will be the responsibility of C code to `free()` it later whenever it wants.

Reply:
no, that's exactly what I want to prevent. Giving away the responsibility of cleaning the memory to the C context will be a security hole that will haunt Prisma for years.